packages/flutter_tools/test/general.shard/android/android_sdk_test.dart

```diff
@@ -342,6 +347,114 @@ void main() {
       Config: () => config,
     });
   });
+
+  group('java version', () {
+    const String exampleJdk8Output = '''
+java version "1.8.0_202"
+Java(TM) SE Runtime Environment (build 1.8.0_202-b10)
+Java HotSpot(TM) 64-Bit Server VM (build 25.202-b10, mixed mode)
+''';
+    // Example strings came from actual terminal output.
+    testWithoutContext('parses jdk 8', () {
+      expect(AndroidSdk.parseJavaVersion(exampleJdk8Output), '1.8.0');
+    });
+
+    testWithoutContext('parses jdk 11 windows', () {
+      const String exampleJdkOutput = '''
+java version "11.0.14"
+Java(TM) SE Runtime Environment (build 11.0.14+10-b13)
+Java HotSpot(TM) 64-Bit Server VM (build 11.0.14+10-b13, mixed mode)
+''';
+      expect(AndroidSdk.parseJavaVersion(exampleJdkOutput), '11.0.14');
+    });
+
+    testWithoutContext('parses jdk 11 mac/linux', () {
+      const String exampleJdkOutput = '''
+openjdk version "11.0.18" 2023-01-17 LTS
+OpenJDK Runtime Environment Zulu11.62+17-CA (build 11.0.18+10-LTS)
+OpenJDK 64-Bit Server VM Zulu11.62+17-CA (build 11.0.18+10-LTS, mixed mode)
+''';
+      expect(AndroidSdk.parseJavaVersion(exampleJdkOutput), '11.0.18');
+    });
+
+    testWithoutContext('parses jdk 17', () {
+      const String exampleJdkOutput = '''
+openjdk 17.0.6 2023-01-17
+OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
+OpenJDK 64-Bit Server VM (build 17.0.6+0-17.0.6b802.4-9586694, mixed mode)
+''';
+      expect(AndroidSdk.parseJavaVersion(exampleJdkOutput), '17.0.6');
+    });
+
+    testWithoutContext('parses jdk 19', () {
+      const String exampleJdkOutput = '''
+openjdk 19.0.2 2023-01-17
+OpenJDK Runtime Environment Homebrew (build 19.0.2)
+OpenJDK 64-Bit Server VM Homebrew (build 19.0.2, mixed mode, sharing)
+''';
+      expect(AndroidSdk.parseJavaVersion(exampleJdkOutput), '19.0.2');
+    });
+
+    testUsingContext('getJavaBinary with AS install', () {
+      final Directory sdkDir = createSdkDirectory(fileSystem: fileSystem);
+      config.setValue('android-sdk', sdkDir.path);
+      final AndroidStudio androidStudio = FakeAndroidStudio();
+
+      final String javaPath = AndroidSdk.findJavaBinary(
+          androidStudio: androidStudio,
+          fileSystem: fileSystem,
+          operatingSystemUtils: FakeOperatingSystemUtils(),
+          platform: platform)!;
+      // Built from the implementation of findJavaBinary android studio case.
+      final String expectedJavaPath = '${androidStudio.javaPath}/bin/java';
+
+      expect(javaPath, expectedJavaPath);
+    }, overrides: <Type, Generator>{
+      FileSystem: () => fileSystem,
+      ProcessManager: () => processManager,
+      Config: () => config,
+      Platform: () => FakePlatform(environment: <String, String>{}),
+    });
+
+    group('java', () {
+      final AndroidStudio androidStudio = FakeAndroidStudio();
```

Comment:
I am not a fan of this pattern and am looking for advice. 

The variable `androidStudio` need to be referenced in `testUsingContext` and in the `overrides` for `testUsingContext. That means it cannot be created in a setUp method. 
So I just create the object but that feels like a hard to debug test pollution issue waiting to happen. 

I think there is knowledge I am missing that covers when to use globals in code and how to override them in test code.

Reply:
You can use `late` and then initialize it within `setUp()`